### DISCUSSION TITLE
能否将包安装到本地并合成一个目录结构

### DISCUSSION THREAD

**Discussion by congcongkuo** [xmake-io/xmake#5431] [Ideas]
生成一个包含 /bin, /lib, /include 的目录，如此 xrepo 可以部分替代系统的包管理器。

xrepo 的官方仓库中有很多工具型的软件 (utilities)，当前这些包的可执行文件被分散在不同的  .xmake/<包名>/bin 目录下，不便于使用。如果能够合并安装，就可以直接将其添加到环境变量 PATH 里。 

[vcpkg](https://learn.microsoft.com/en-us/vcpkg/consume/manifest-mode) 有此功能。

**Comment by star-hengxing** [comment]
这种？
https://xrepo.xmake.io/#/getting_started?id=package-virtual-environment
https://tboox.org/cn/2022/03/07/xmake-update-v2.6.4

**Reply by congcongkuo** [reply]
虚拟环境是把各个 package 的 /bin, /lib, /include 等子目录依次加到 PATH, LD_LIBRARY_PATH, CPATH 等环境变量里面，package 依然是分散的。

虚拟环境有时候不如构建成一个目录的模式灵活，比如我希望在 shell 中可以直接调用 xrepo 安装的软件就需要逐项把 /bin 的路径加到 shell 的 rc 文件中。给代码分析软件添加 include path，也需要逐个 package 添加 include path。

而且因为 package 的路径中包含版本号，管理起来不是很方便。如果能分成一个本地目录的话，就可以很方便的进行更新。目录的路径保持不变，配置文件中的路径依然有效，但是软件已经得到了更新。其实就是 linux 中系统包管理器的工作。

以下是别的一些工具的做法：

在 vcpkg 中软件是 vcpkg/buildtrees 目录下构建的，然后被复制到了当前工作目录。缺陷就是有些 package 是不能移动位置的，vcpkg 的作法是不收录不能移动的 package.

nix 的做法是真正的 package 全部在 /nix 目录下，本地的目录确实是一堆软链接。当然 nix 也支持虚拟环境。缺陷就是创建 /nix 在一些系统上需要特别的处理，以及很难支持 windows。

我觉得创建软链接的方式或许可供参考。

**Comment by waruqi** [comment]
> 虚拟环境有时候不如构建成一个目录的模式灵活，比如我希望在 shell 中可以直接调用 xrepo 安装的软件就需要逐项把 /bin 的路径加到 shell 的 rc 文件中。给代码分析软件添加 include path，也需要逐个 package 添加 include path。

不需要，直接 xmake.lua `add_requires("luajit", "python 3.x", "llvm")` 配置上各种需要的包，然后执行 `xrepo env shell` 进入环境，就可以直接全部使用了，你先怎么组合随你，都虚拟环境，怎么可能还会让你自己去挨个 includes 。。具体用法看下文档，另外也有一些内置常用的环境，可以直接加载进入使用。。

> nix 的做法是真正的 package 全部在 /nix 目录下，本地的目录确实是一堆软链接。当然 nix 也支持虚拟环境。缺陷就是创建 /nix 在一些系统上需要特别的处理，以及很难支持 windows。
> 我觉得创建软链接的方式或许可供参考。

不考虑，软链的方式，反而对 windows 上不友好，win 上软链出了名的难用，还要管理员权限，基本玩不起来，但是这边要做就是要全平台的，不可能排除 win

另外，软链还有个问题，就是各种版本间的软链冲突。。比如一个程序需要 /bin/python 是 2.0 的，另外一个需要 /bin/python 是 3.0 ，又需要同时运行，我怎么链？这只是个例子，还要各种 lib 冲突，资源文件冲突，想要处理好不容易。。但是走虚拟环境，各种版本组合 你随意快速切换，想怎么用 就怎么用。。也不用为此改名 python2 python3
